gnocchi/storage/__init__.py

```diff
@@ -28,9 +28,14 @@
                default='file',
                help='Storage driver to use'),
     cfg.IntOpt('metric_processing_delay',
-               default=60,
+               default=30,
                help="How many seconds to wait between "
                "scheduling new metrics to process"),
+    cfg.IntOpt('tasks_per_worker',
+               default=256,
+               min=1,
+               help="How many tasks to assign each metricd worker when "
+               "scheduling measures processing jobs"),
```

Comment:
i think this is only really useful on ceph tbh. should probably add that to help?

Reply:
Not really, it's useful for any driver that wants to tweak the throughput.

Reply:
that's kinda my point. the throughput is some combination of: processing_delay, number of unqiue metrics, metrics posting interval, and number of workers.

that's the reason for my concern with setting default to 256, it only makes sense for specific scenario. we need to clarify what that scenario is because TASKS_PER_WORKER = 16 worked  fine for my scenario (1000metrics, every minute, 72 workers).
e for my scenario (1000metrics,

Reply:
72 workers is *huge* for 1k metrics which is *small*.

This is exactly why we need to change to 256. You are fixing the CPU not being used at 100% per worker over a your interval by running A LOT of workers which consumes a ton of memory. It's way more memory friendly to run less workers actually doing a lot of their jobs.

Ultimately I don't mind leaving 16 as a default. We'll carry a different default in TripleO, that's all :/

Reply:
sorry, i meant 1K resources and 20K metrics (still small)... but this is with ceph. the impact this has on the non-ceph drivers is different.

point is, i  think we need to define what the default is configured for, not that 256 (or any number) is wrong.

Reply:
Well the current default of 60s for `metric_processing_delay` and 16 `tasks_per_worker` *cannot* keep a worker busy for more than 3 seconds unless you're using an old Pentium 133 MHz.
So even we switch to a default of 5s for `metric_processing_delay`, which will heavily hit the the storage by listing over and over again the measures, or we keep the CPU busy for a decent amount of time (at least `metric_processing_delay` seconds) which means grabbing a whole bunch of measures to churn on, i.e. 256.

This is not perfect, and this algorithm cannot be perfect and it needs tuning based on your workload, anyway. But I can't see how the current default are ok on any decent computer. :(

Reply:
again

> point is, i think we need to define what the default is configured for, not that 256 (or any number) is wrong.

so if the default is designed for 30s delay, x metrics, y metricd, etc... which should note it. or else the user has no idea what they should change for their system.

Reply:
I don't think you understand the problem completely.
The default (workers + delay + tasks_per_worker) are designed for processors < 200 MHz or a workload of `16 * workers` metrics. That's ridiculous.

We can document that, for sure. But it seems counter-productive to have defaults that are that shitty… so what about we fix them in the meantime? That's what this patch's trying to do. :)

Reply:
i know the problem. it's why i always tell people to change TASKS_PER_WORKER :p 

what i'm suggesting is we document better as only people who know the scheduling works in v3.x can actually figure out that this value should be adjusted. something to inform user that TASKS_PER_WORKER * workers should equal the number of metrics (or measures for ceph case) accumulated over processing_delay.